File: marketing-worker/src/main/java/com/marketing/worker/service/EvolutionService.java
package com.marketing.worker.service;

import com.marketing.worker.persistence.entity.CampaignEntity;
import com.marketing.worker.persistence.entity.ChannelConfig;
import com.marketing.worker.persistence.entity.ContactEntity;
import com.marketing.worker.persistence.repository.ChannelConfigRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import java.time.Duration;
import java.util.HashMap;
import java.util.Map;

@Service
@RequiredArgsConstructor
@Slf4j
public class EvolutionService {

    private final WebClient.Builder webClientBuilder;
    private final ChannelConfigRepository channelConfigRepository;

    @Value("${evolution.api.url}")
    private String apiUrl;

    @Value("${evolution.api.key}")
    private String globalApiKey;

    public Mono<String> sendMessage(CampaignEntity campaign, ContactEntity contact, String formattedMessage) {
        return channelConfigRepository.findById(campaign.getChannelId())
                .flatMap(config -> {
                    String phone = contact.getPhone().replaceAll("[^0-9]", "");
                    String apiKey = config.getApiKey() != null ? config.getApiKey() : globalApiKey;

                    // 1. Simulate "composing" presence before sending
                    return sendPresence(config.getInstanceName(), phone, apiKey)
                            .then(Mono.delay(Duration.ofMillis(1500 + (long)(Math.random() * 2000))))
                            .then(Mono.defer(() -> {
                                if (campaign.getMediaUrl() != null && !campaign.getMediaUrl().isEmpty()) {
                                    return sendMedia(config, phone, campaign.getMediaUrl(), campaign.getMediaType(), formattedMessage, apiKey);
                                } else {
                                    return sendText(config, phone, formattedMessage, apiKey);
                                }
                            }));
                });
    }

    /**
     * Sends a "composing" presence indicator so WhatsApp sees
     * the number as "typing..." before the actual message arrives.
     */
    private Mono<Void> sendPresence(String instanceName, String phone, String apiKey) {
        String url = apiUrl + "/chat/sendPresence/" + instanceName;
        Map<String, Object> body = new HashMap<>();
        body.put("number", phone);
        body.put("presence", "composing");
        body.put("delay", 5000);

        return webClientBuilder.build().post()
                .uri(url)
                .header("apikey", apiKey)
                .contentType(MediaType.APPLICATION_JSON)
                .bodyValue(body)
                .retrieve()
                .bodyToMono(Void.class)
                .onErrorResume(e -> {
                    log.debug("Presence update skipped for {}: {}", phone, e.getMessage());
                    return Mono.empty();
                });
    }

    private Mono<String> sendText(ChannelConfig config, String phone, String text, String apiKey) {
        String url = apiUrl + "/message/sendText/" + config.getInstanceName();
        Map<String, Object> body = new HashMap<>();
        body.put("number", phone);
        body.put("text", text);
        body.put("delay", 1200 + (int)(Math.random() * 3000));

        return webClientBuilder.build().post()
                .uri(url)
                .header("apikey", apiKey)
                .contentType(MediaType.APPLICATION_JSON)
                .bodyValue(body)
                .retrieve()
                .bodyToMono(Map.class)
                .map(response -> {
                    log.debug("✅ Message sent to {}", phone);
                    return extractMessageId(response);
                })
                .onErrorResume(e -> {
                    log.error("❌ Error sending message to {}: {}", phone, e.getMessage());
                    return Mono.error(e);
                });
    }

    private String extractMessageId(Map response) {
        try {
            if (response.containsKey("key")) {
                Map key = (Map) response.get("key");
                return (String) key.get("id");
            }
        } catch (Exception e) {
            log.warn("Could not extract message ID from response: {}", e.getMessage());
        }
        return null;
    }

    private Mono<String> sendMedia(ChannelConfig config, String phone, String mediaUrl, String mediaType, String caption, String apiKey) {
        String url = apiUrl + "/message/sendMedia/" + config.getInstanceName();
        Map<String, Object> body = new HashMap<>();
        body.put("number", phone);
        body.put("media", mediaUrl);
        body.put("mediatype", mediaType != null ? mediaType.toLowerCase() : "image");
        body.put("caption", caption);
        body.put("delay", 1500 + (int)(Math.random() * 3000));

        return webClientBuilder.build().post()
                .uri(url)
                .header("apikey", apiKey)
                .contentType(MediaType.APPLICATION_JSON)
                .bodyValue(body)
                .retrieve()
                .bodyToMono(Map.class)
                .map(response -> {
                    log.debug("✅ Media sent to {}", phone);
                    return extractMessageId(response);
                })
                .onErrorResume(e -> {
                    log.error("❌ Error sending media to {}: {}", phone, e.getMessage());
                    return Mono.error(e);
                });
    }
}

File: marketing_agent/services/evolution_service.py
import time, random, logging, requests
from config import Config

logger = logging.getLogger(__name__)

class EvolutionService:
    def __init__(self):
        self.api_url = Config.EVOLUTION_API_URL
        self.api_key = Config.EVOLUTION_API_KEY
        self.instance = Config.EVOLUTION_INSTANCE
        self.headers = {"apikey": self.api_key, "Content-Type": "application/json"}

    def send_campaign(self, phone: str, message: 'CampaignMessage') -> bool:
        try:
            self._send_presence(phone, "composing")
            time.sleep(1.5 + random.random() * 2.0)
            if message.media_url:
                return self._send_media(phone, message)
            else:
                return self._send_text(phone, message.text)
        except Exception as e:
            logger.error(f"Error sending message to {phone}: {e}")
            return False

    def _send_presence(self, phone: str, presence: str):
        url = f"{self.api_url}/chat/sendPresence/{self.instance}"
        payload = {"number": phone, "presence": presence, "delay": 5000}
        try:
            requests.post(url, json=payload, headers=self.headers, timeout=10)
        except Exception:
            pass

    def _send_text(self, phone: str, text: str) -> bool:
        url = f"{self.api_url}/message/sendText/{self.instance}"
        payload = {"number": phone, "text": text, "delay": 1200 + random.randint(0, 3000)}
        resp = requests.post(url, json=payload, headers=self.headers, timeout=30)
        resp.raise_for_status()
        logger.info(f"✅ Text sent to {phone}")
        return True

    def _send_media(self, phone: str, message: 'CampaignMessage') -> bool:
        url = f"{self.api_url}/message/sendMedia/{self.instance}"
        payload = {
            "number": phone,
            "media": message.media_url,
            "mediatype": message.media_type or "image",
            "caption": message.caption or message.text,
            "delay": 1500 + random.randint(0, 3000),
        }
        resp = requests.post(url, json=payload, headers=self.headers, timeout=30)
        resp.raise_for_status()
        logger.info(f"✅ Media sent to {phone}")
        return True
